internals/metrics/metrics.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package metrics
+
+import (
+	"fmt"
+	"io"
+)
+
+type MetricType int
+
+const (
+	TypeCounterInt MetricType = iota
```

Comment:
It's probably worth doing `iota+1` so these start at one, so we don't accidentally forget to set `Type` and get the zero value (`TypeCounterInt`) by accident.